### PR TITLE
Integrate node-problem-detector

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**: This PR integrates node-problem-detector in Gardener as a shoot-core addon. The chart is inspired by the following references:
- https://github.com/helm/charts/tree/master/stable/node-problem-detector
- https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/node-problem-detector

**Which issue(s) this PR fixes**:
Fixes #[312](https://github.com/gardener/machine-controller-manager/issues/312)

**Special notes for your reviewer**: As it's a new component, please have a special eye on labels and other metadata to maintain homogeneity.

**Release note**:

```improvement user
Node-problem-detector is integrated into Gardener. Certain problems occured in nodes will be visible as part of `NodeConditions` or `Event`. Refer [NPD](https://github.com/kubernetes/node-problem-detector) doc for more info.
```

### CODE REVIEW

File: charts/shoot-core/components/charts/node-problem-detector/templates/daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+apiVersion: apps/v1
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/node-problem-detector/templates/daemonset.yaml]:
```suggestion
apiVersion: {{ include "daemonsetversion" . }}
```

File: charts/shoot-core/components/charts/node-problem-detector/templates/clusterrolebinding.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.rbac.create -}}
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/node-problem-detector/templates/clusterrolebinding.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/shoot-core/components/charts/node-problem-detector/templates/clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{{- if .Values.rbac.create -}}
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/node-problem-detector/templates/clusterrole.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

**Review by rfranzke — COMMENTED** [review]
Looks good so far, thanks! Why is it still "WIP"?
Also, did you test with setting `.spec.kubernetes.allowPrivilegedContainers=false`? Does NPD work properly in this case? Or are additional PSPs needed?

File: charts/shoot-core/components/charts/node-problem-detector/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
+name: node-problem-detector
+version: "0.7.1"
```
Comment (swapnilgm) [inline charts/shoot-core/components/charts/node-problem-detector/Chart.yaml]: Shoudn't it be "1.6.1" as per https://github.com/helm/charts/blob/master/stable/node-problem-detector/Chart.yaml#L3
```suggestion
version: "1.6.1"
```

File: pkg/operation/common/types.go [hunk 5]
```diff
@@ -259,6 +259,9 @@ const (
 	// KubeProxyDaemonSetName is the name of the kube-proxy daemon set.
 	KubeProxyDaemonSetName = "kube-proxy"
 
+	// NodeProblemDetectorDaemonSetName is the name of the kube-proxy daemon set.
```
Comment (swapnilgm) [inline pkg/operation/common/types.go]:
```suggestion
	// NodeProblemDetectorDaemonSetName is the name of the node-problem-detector daemon set.
```

File: pkg/operation/common/types.go [hunk 6]
```diff
@@ -418,6 +421,9 @@ const (
 	// CoreDNSImageName is the name of the CoreDNS image.
 	CoreDNSImageName = "coredns"
 
+	// NodeProblemDetectorImageName is the name of the CoreDNS image.
```
Comment (swapnilgm) [inline pkg/operation/common/types.go]:
```suggestion
	// NodeProblemDetectorImageName is the name of the node-problem-detector image.
```

**Review by swapnilgm — COMMENTED** [review]
Instead of duplicating and maintaining NPD chart on  our own, how about managing NPD dependency by adding `requirements.yaml` to [componnets chart](https://github.com/gardener/gardener/tree/master/charts/shoot-core/components) And adding section for NPD under `values.yaml`.
Ref: https://helm.sh/docs/developing_charts/#chart-dependencies

Only thing stopping from doing this is the core resources APIVersion configuration. As i understood most of stable help chart don't allow configurable APIVersion for different resource and gardener does use different APIVersion for core resource based on k8s version. But if we consider external stable charts they are applicable to certain k8s version only. And that can be coded easily in gardener to allow them for those specific k8s versions only without having configurable APIVersion for core resources. So, I think we should stop maintaining external components charts with modification.
@rfranzke WDYS?

@hardikdr Why there is difference between the chart you added and upstream stable chart. Ideally the change should be only with `values.yaml`.

**Comment by ialidzhikov** [comment]
Is the node-problem-detector installed on each Shoot? What's the motivation behind it? I guess the idea was to keep addons out of gardener. g/bouquet or something else? Is it special system addon that we want to install always?

/cc @mvladev

**Comment by dguendisch** [comment]
@ialidzhikov the motivation is that there are several circumstances that render your node useless/broken without the kubelet detecting this. We ran several times into corrupt and thus read-only filesystems (the node is thus broken) but scheduler and kubelet happily trying to start one pod after the other on that broken node. Such situations needed manual interaction because of the missing problem detection.
Hence I actually cannot imagine that one wants to run a cluster without NPD :) (except however when you want to deploy/tune your own NPD).

@hardikdr: If I get it right, this PR is only about "detecting" problems, not about "acting" on them, correct? So while this will for example detect a read-only filesystem, mcm won't automagically react on it, or is there already some mcm PR to react on the problems reported by NPD?

**Comment by hardikdr** [comment]
@rfranzke thanks for reviewing.
>Why is it still "WIP"? 

I was testing the PR locally for a few cases, so I kept WIP. Removing now, as it's done.

>did you test with setting .spec.kubernetes.allowPrivilegedContainers=false?

I haven't really, but as NPD needs to interact with system often, I assumed it'd need to be privileged-container. Though a valid point, I will quickly cross-check and update the thread.

**Comment by hardikdr** [comment]
@swapnilgm thanks for reviewing.
>Instead of duplicating and maintaining NPD chart on our own, how about managing NPD dependency by adding requirements.yaml

That's a good point, we could adopt that for all the charts we consume directly from upstream, I'd generally +1 on not maintaining the code/charts separately wherever it is possible. 
Although in this case, on top of `APIVersion` conflict you mentioned, there are few minor changes I have inside the chart as well. 
Eg. In [daemonset.yaml](https://github.com/gardener/gardener/blob/bffcf6eee4b07c63dbe6137e9f27480c6d10ca78/charts/shoot-core/components/charts/node-problem-detector/templates/daemonset.yaml#L41-L48) 
- Updating the exec command to include more monitors, using gardener friendly labels/naming-conventions and so on. 

Overall I am open for suggestions and happy to fetch chart as it, with few customizations on top as mentioned above.

**Comment by ialidzhikov** [comment]
> Hence I actually cannot imagine that one wants to run a cluster without NPD :) (except however when you want to deploy/tune your own NPD).

Then I would expect a way to disable it from `spec.addons`. I still miss the whole picture. NPD adds new NodeConditions and MCM starts watching on them?

**Comment by rfranzke** [comment]
> Instead of duplicating and maintaining NPD chart on our own, how about managing NPD dependency by adding requirements.yaml to componnets chart And adding section for NPD under values.yaml.

I don't think our chart rendering mechanism supports this.

> Then I would expect a way to disable it from spec.addons. I still miss the whole picture. NPD add new NodeConditions and MCM starts watching on them?

I don't think we should allow end-users to configure it. It comes automatically with every shoot as it helps us providing a better service quality (as we now can more easily detect broken nodes and replace them). For example, the node-exporter daemonset can't be deactivated either by end-users because it's required as part of our monitoring offering.

**Comment by dguendisch** [comment]
> I still miss the whole picture. NPD adds new NodeConditions and MCM starts watching on them?

Exactly!
(Side note: deploying NDP on my own is kind of hacky, because it only reports, but so far no one reacts on the problems found, hence today I have NPD deployed +an additional script in case NPD problems are found that "kills" stuff on the node, so that ultimately kubelet detects it and mcm can react...)

**Comment by hardikdr** [comment]
>If I get it right, this PR is only about "detecting" problems, not about "acting" on them, correct? So while this will for example detect a read-only filesystem, mcm won't automagically react on it, or is there already some mcm PR to react on the problems reported by NPD?

Yes, that's true, we will soon be opening a PR on MCM to act on a very minimal set of `NodeConditions`. For now, it may be `ReadonlyFilesystem ` and `KernelDeadlock`. MCM will simply replace machines with such NodeConditions set to true after a configurable timeout. As when we see a need, we may enable MCM to react on more NodeConditions or integrate remedy-systems to react on them.

**Comment by swapnilgm** [comment]
> > Instead of duplicating and maintaining NPD chart on our own, how about managing NPD dependency by adding requirements.yaml to componnets chart And adding section for NPD under values.yaml.
> 
> I don't think our chart rendering mechanism supports this.

i was under the assumption that  our chart rendering library is wrapper over helm chart rendering so it should have supported it. Then does it makes sense to add that support in our chart rendering lib? 

This way we can also force our other components like MCM, etcd-backup-restore, gardener-resource-manager, terraformer to maintain helm chart on it. and reuse it in gardener. Single point . of maintenance. Same is the case for landscape/garden-setup.

**Comment by rfranzke** [comment]
> i was under the assumption that our chart rendering library is wrapper over helm chart rendering so it should have supported it. Then does it makes sense to add that support in our chart rendering lib?

Try it out, I'm not sure whether it works.

> This way we can also force our other components like MCM, etcd-backup-restore, gardener-resource-manager, terraformer to maintain helm chart on it. and reuse it in gardener. Single point . of maintenance. Same is the case for landscape/garden-setup.

 Yes, this would be a nice benefit, however, on the other hand I think at least some of the charts are specific to the way how gardener deploys them. Also, this also introduces a new dependency to these charts, i.e., if there is a bug/problem with such a chart then it has to first be fixed, released, and then updated in this repo. Kind of "not nice", I guess. Definitely, whatever we plan here should happen separately and is out of scope of NPD integration.

**Comment by swapnilgm** [comment]
> > i was under the assumption that our chart rendering library is wrapper over helm chart rendering so it should have supported it. Then does it makes sense to add that support in our chart rendering lib?
> 
> Try it out, I'm not sure whether it works.
> 
> > This way we can also force our other components like MCM, etcd-backup-restore, gardener-resource-manager, terraformer to maintain helm chart on it. and reuse it in gardener. Single point . of maintenance. Same is the case for landscape/garden-setup.
> 
> Yes, this would be a nice benefit, however, on the other hand I think at least some of the charts are specific to the way how gardener deploys them. Also, this also introduces a new dependency to these charts, i.e., if there is a bug/problem with such a chart then it has to first be fixed, released, and then updated in this repo. Kind of "not nice", I guess. Definitely, whatever we plan here should happen separately and is out of scope of NPD integration.

Agreed. Sorry for diverging topic from NPD.

**Comment by hardikdr** [comment]
@ialidzhikov Overall I agree with @rfranzke's comment above, we basically want to deliberately enable the NPD for all shoots, considering it adds a core service quality of `identifying` the problems at the node. Though we might want to think of making the impact of `future-remedy-systems` optional if the user would want to opt-out of certain/unusual cases, but that's far fetched.

**Review by rfranzke — DISMISSED** [review]
/lgtm, please squash the commits.

**Review by rfranzke — DISMISSED** [review]

**Comment by rfranzke** [comment]
@hardikdr mentioned that NPD does not yet work for shoots if it is configured in a way that privileged containers are not allowed. /cc @mvladev

**Review by swapnilgm — DISMISSED** [review]
LGTM/

**Comment by hardikdr** [comment]
As discussed with @mvladev, we'll introduce the necessary PSPs from the upstream chart with minor modifications to allow the privileged containers.

PS: I'm off tomorrow, will try to fix this early next week.

**Comment by vpnachev** [comment]
> Yes, that's true, we will soon be opening a PR on MCM to act on a very minimal set of `NodeConditions`. For now, it may be `ReadonlyFilesystem ` and `KernelDeadlock`. MCM will simply replace machines with such NodeConditions set to true after a configurable timeout.

This will require something like `MachineDisruptionBudget`, otherwise it can pull off all of the machines...

**Comment by rfranzke** [comment]
@hardikdr are PDBs respected in case multiple machines are marked with these conditions by NPD and MCM decides to take them off?

**Comment by hardikdr** [comment]
> This will require something like `MachineDisruptionBudget`, otherwise it can pull off all of the machines...

Yes, that's a good point. We'd eventually need to introduce something similar. I would though prefer to introduce a mechanism of freezing the machineset/machinedeployment when X% [50/60] of the machines are unhealthy simultaneously.  This would indicate more of an infra-failure than actual issues with machines.  
   - Otherwise, if machines are permanently damaged due to any of `KuberneteReady`, `KernelPanic` or `ReadOnlyFilesystem` since the timeout period, it's very likely that workload is also impacted and preserving such machines might not be useful anyways. 

> @hardikdr are PDBs respected in case multiple machines are marked with these conditions by NPD and MCM decides to take them off?

The short answer is yes, but there is more to it. We always drain the machines before deleting them, which inherently ensures that PDBs are respected by nature. Basically drain calls `EvictionAPI` for each pod on a machine, which avoids deleting the pods with PDBs-violated. Though we also make sure, that such machines are not forever stuck due to misuse of PDBs and delete them after a timeout period.

**Comment by hardikdr** [comment]
Tested with `.spec.kubernetes.allowPrivilegedContainers=false` works well now.

**Review by rfranzke — APPROVED** [review]

**Review by zanetworker — APPROVED** [review]

**Pull request merged**